clock/clock.go

```diff
@@ -0,0 +1,23 @@
+// Package clock is a library for testing time related code.  It exports an interface Clock that
+// mimics the standard library time package functions.  In production, an implementation that calls
+// thru to the standard libary is used.  In testing, a Mock clock is used to precisely control and
+// intercept time functions.
+package clock
+
+import (
+	"context"
+	"time"
+)
+
+type Clock interface {
+	// NewContextTicker is a convenience function that calls f on the interval d until either the
+	// given context expires or f returns an error.  Callers may call Wait() on the returned Waiter
+	// to wait until this happens and obtain the errror.
+	NewContextTicker(ctx context.Context, d time.Duration, f func() error, tags ...string) Waiter
```

Comment:
How would you feel about using functional options instead of `...string`? An alternative to having this convenience function would be to accept the following signature:

`NewTicker(d time.Duration, f func() error, opts ...Option)` and you'd pass it `clock.WithContext(ctx)` or `clock.Tags("tag1", "tag2") as options.

I think this could result in a slightly cleaner and more flexible API, as well as the added verbosity/explicitness can help with code reads (i.e. "what are these strings?").

In this scenario, I imagine the return signature of `NewTicker` could include all three methods `Stop`, `Reset` and _`Wait`_.

Reply:
Firstly, maybe "ContextTicker" isn't the best name.  It's not a ticker that optionally takes a context, it's a ticker calls the function `f` on the ticks, rather than just sending the ticks to a channel.  Sort of like `AfterFunc` is a special timer that calls a function.  "TickerFunc"?

Secondly, the idea is to include tags on _every_ call in the `clock` interface, and all the timer/ticker calls as well.  With the exception of this ContextTicker/TickerFunc, the idea is to verbatim replicate the functionality of the `time` package, not enhance it. So, I literally cannot think of another option we'd want other than tags, and even if I could, it's unlikely to apply to _every_ call, so adding it to a generic `Option` interface would muddy the waters. I think variability is actually a downside in this case, and we should keep it simple. 

Lastly, I'm not a fan of the verbosity of `clock.Tags("tag1")`.  The signature says `tags ...string`, so I feel like that's clear enough once you jump to the definition.